src/middleware/compression.rs

```diff
@@ -0,0 +1,420 @@
+                    let stream = stream::BrotliDecoder::new(body);
+                    *req.body_mut() = Body::from_stream(stream);
+                }
+                Encoding::Zstd => {
+                    let mut body = Body::empty();
+                    std::mem::swap(&mut body, req.body_mut());
+                    let stream = stream::ZstdDecoder::new(body);
+                    *req.body_mut() = Body::from_stream(stream);
+                }
+                Encoding::Identity => (),
+            }
+        }
+
+        // strip the content-encoding header
+        req.headers_mut().remove(CONTENT_ENCODING).unwrap();
+
+        Ok(())
+    }
+}
+
+impl<Data: Send + Sync + 'static> Middleware<Data> for Decompression {
+    fn handle<'a>(
+        &'a self,
+        mut cx: Context<Data>,
+        next: Next<'a, Data>,
+    ) -> BoxFuture<'a, Response> {
+        box_async! {
+            match self.decode(cx.request_mut()) {
+                Ok(_) => (),
+                Err(e) => return e.into_response(),
+            };
+            next.run(cx).await
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use async_compression::flate2;
+    use bytes::Bytes;
+    use futures::{
+        executor::{block_on, block_on_stream},
+        stream::StreamExt,
+    };
+    use http::header::ACCEPT_ENCODING;
+    use http_service::Body;
+    use http_service_mock::make_server;
+
+    async fn lorem_ipsum(_cx: Context<()>) -> String {
+        String::from(r#"
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam rutrum et risus sed egestas. Maecenas dapibus enim a posuere
+            semper. Cras venenatis et turpis quis aliquam. Suspendisse eget risus in libero tristique consectetur. Ut ut risus cursus, scelerisque
+            enim ac, tempus tellus. Vestibulum ac porta felis. Aenean fringilla posuere felis, in blandit enim tristique ut. Sed elementum iaculis
+            enim eu commodo.
+        "#)
+    }
+
+    fn lorem_ipsum_bytes() -> Vec<u8> {
+        String::from(r#"
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam rutrum et risus sed egestas. Maecenas dapibus enim a posuere
+            semper. Cras venenatis et turpis quis aliquam. Suspendisse eget risus in libero tristique consectetur. Ut ut risus cursus, scelerisque
+            enim ac, tempus tellus. Vestibulum ac porta felis. Aenean fringilla posuere felis, in blandit enim tristique ut. Sed elementum iaculis
+            enim eu commodo.
+        "#).into_bytes()
+    }
+
+    // Echoes the request body in bytes.
+    async fn echo_bytes(mut cx: Context<()>) -> Vec<u8> {
+        cx.body_bytes().await.unwrap()
+    }
+
+    // Generates the app.
+    fn app() -> crate::App<()> {
+        let mut app = crate::App::new();
+        app.at("/").get(lorem_ipsum);
+        app.at("/echo").get(echo_bytes);
+        app.middleware(Compression::new());
+        app.middleware(Decompression::new());
```

Comment:
Should we make it possible for people to use only one of `Compression` and `Decompression`?

Reply:
I hadn't checked if limiting users to only one use of Compression and Decompression middleware each was possible yet, actually. Though I don't believe it can be done in this file -- such a check would likely exist in `src/middleware/mod.rs` or something, I imagine...

Reply:
Oh, I will remark though -- including multiple `Decompression` middlewares will usually have no side effect (aside from going through an *extra* middleware that doesn't do anything, as the decompression middleware already strips out the `Content-Encoding` header on finishing the decompression.

Including multiple `Compression` middlewares... likely will. I don't think users will willingly do this, but if there is a way to warn them of duplicate middlewares I'd be for it.

Reply:
I wonder if there are any ways to warn users when multiple `Compression` or `Decompression` is used.

Reply:
It might be worth the `Compression` middleware checking for an existing `Content-Encoding` header on the body to avoid double encoding (e.g. it's relatively common to have static file servers that check for existing pre-compressed versions of files and serve those based on `Accept-Encoding`, you should be able to use both that and a generic `Compression` middleware to compress other responses that weren't pre-compressed).

Reply:
Oh, good catch. I can probably throw that in real quick.